src/chabochethermal.jl

```diff
@@ -0,0 +1,1050 @@
+                + dcontract(dDdthetaf(temperature_new), elastic_strain) * dtemperature)
+    end
+
+    stress += elastic_dstress(dstrain, dtemperature)
+
+    # using elastic trial problem state
+    if ff(stress, R, X1, X2, X3, temperature) > 0.0  # plastic region
+        rx!, rdstrain, rtemperature = create_nonlinear_system_of_equations(material)
+        x0 = state_to_vector(stress, R, X1, X2, X3)
+        res = nlsolve(rx!, x0; method=material.options.nlsolve_method, autodiff=:forward)  # user manual: https://github.com/JuliaNLSolvers/NLsolve.jl
+        converged(res) || error("Nonlinear system of equations did not converge!")
+        x = res.zero
+        stress, R, X1, X2, X3 = state_from_vector(x)
+
+        # using the new problem state
+        temperature_new = temperature + dtemperature
+
+        # Compute the new plastic strain
+        dotp = dotpf(stress, R, X1, X2, X3, temperature_new)
+        n = nf(stress, R, X1, X2, X3, temperature_new)
+
+        dp = dotp * dtime  # Δp, using backward Euler (dotp is |∂εpl/∂t| at the end of the timestep)
+        plastic_strain += dp * n
+        cumeq += dp   # cumulative equivalent plastic strain (note Δp ≥ 0)
+
+        # Compute the new algorithmic jacobian Jstrain by implicit differentiation of the residual function,
+        # using `ForwardDiff` to compute the derivatives. Details in `create_nonlinear_system_of_equations`.
+        # We compute ∂V/∂D ∀ V ∈ state, D ∈ drivers (excluding time).
+        drdx = ForwardDiff.jacobian(debang(rx!), x)
+
+        # Here we don't bother with offdiagscale, since this Voigt conversion is just a marshaling.
+        # All `rdstrain` does with the Voigt `dstrain` is to unmarshal it back into a tensor.
+        # All computations are performed in tensor format.
+        rdstrainf(dstrain) = rdstrain(stress, R, X1, X2, X3, dstrain)  # at solution point
+        drdstrain = ForwardDiff.jacobian(rdstrainf, tovoigt(dstrain))
+        Jstrain = -drdx \ drdstrain
+        jacobian = fromvoigt(Symm4, Jstrain[1:6, 1:6])
+        dRdstrain = fromvoigt(Symm2, Jstrain[7, 1:6])
+        dX1dstrain = fromvoigt(Symm4, Jstrain[8:13, 1:6])
+        dX2dstrain = fromvoigt(Symm4, Jstrain[14:19, 1:6])
+        dX3dstrain = fromvoigt(Symm4, Jstrain[20:25, 1:6])
+
+        rtemperaturef(theta) = rtemperature(stress, R, X1, X2, X3, theta)  # at solution point
+        drdtemperature = ForwardDiff.jacobian(rtemperaturef, [temperature_new])
+        Jtemperature = -drdx \ drdtemperature
+        dstressdtemperature = fromvoigt(Symm2, Jtemperature[1:6, 1])
+        dRdtemperature = Jtemperature[7, 1]
+        dX1dtemperature = fromvoigt(Symm2, Jtemperature[8:13, 1])
+        dX2dtemperature = fromvoigt(Symm2, Jtemperature[14:19, 1])
+        dX3dtemperature = fromvoigt(Symm2, Jtemperature[20:25, 1])
+    else  # elastic region
+        jacobian = gradient(((dstrain) -> elastic_dstress(dstrain, dtemperature)),
```

Comment:
I like this, as a move towards more generic elasticity definitions.

Reply:
It's also the DRY way to obtain the algorithmic jacobian in the elastic region. :)